File: 16-3sum-closest/16-3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        ans = 0 
        proximity = float('inf')
        nums.sort()
        for i, num1 in enumerate(nums):
            left, right = i+1, len(nums)-1
            while left < right:
                res = nums[left] + nums[right] +num1
                if abs(res-target) < proximity:
                    proximity, ans = abs(res-target), res
                if res > target:
                    right -= 1 
                elif res < target:
                    left += 1
                else:
                    break
        return ans

File: 16-3sum-closest/16-3sum-closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int n = nums.size();
        sort(nums.begin(), nums.end());
        int ans; 
        int proximity = INT_MAX;
          
        
          for(int i=0;i<n-2;i++)
          {
              int left = i+1;
              int right = n-1 ;
              
              while (left < right){
                  int sum = nums[i] + nums[left] + nums[right];
                  if (abs(sum-target) < proximity){
                      proximity = abs(sum-target);
                      ans = sum;
                  }
                  if  (sum > target){
                      right -= 1;
                  }else if (sum < target) {
                      left += 1;
                  }else{
                      return target;
                  }
                  
              }
          }
        
        return ans;
    } 
    
};